### PR TITLE
builder: fix empty reference

### LINKED ISSUE #43653: [regression in master] docker build without --tag tries to tag as empty image name (BuildKit)
While running internal tests against builds from master, I noticed some tests failing; when trying to build an image without explicitly tagging it (`-t` / `--tag`), BuildKit attempts to tag the image with an empty name, causing  an `invalid reference format` error.

All cases below use the same version of the CLI and buildx (I'm running the `docker:20.10-dind` image with the socket mounted)

```bash
docker --version
Docker version 20.10.16, build aa7e414
```

Note that the example uses a remote source, but I've seen other tests in our internal tests fail with the same (building from a Dockerfile)

On a docker 20.10 daemon with `DOCKER_BUILDKIT=1` (not using buildx);

```bash
DOCKER_BUILDKIT=1 docker build https://github.com/docker/docker-bench-security.git#v1.3.5
[+] Building 14.5s (9/9) FINISHED
 => [internal] load git source https://github.com/docker/docker-bench-security.git#v1.3.5                                                                                                              1.5s
 => [internal] load metadata for docker.io/library/alpine:3.10                                                                                                                                         3.7s
 => [auth] library/alpine:pull token for registry-1.docker.io                                                                                                                                          0.0s
 => [1/5] FROM docker.io/library/alpine:3.10@sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98                                                                                   0.8s
 => => resolve docker.io/library/alpine:3.10@sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98                                                                                   0.0s
 => => sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98 1.64kB / 1.64kB                                                                                                         0.0s
 => => sha256:e515aad2ed234a5072c4d2ef86a1cb77d5bfe4b11aa865d9214875734c4eeb3c 528B / 528B                                                                                                             0.0s
 => => sha256:e7b300aee9f9bf3433d32bc9305bfdd22183beb59d933b48d77ab56ba53a197a 1.47kB / 1.47kB                                                                                                         0.0s
 => => sha256:396c31837116ac290458afcb928f68b6cc1c7bdd6963fc72f52f365a2a89c1b5 2.80MB / 2.80MB                                                                                                         0.5s
 => => extracting sha256:396c31837116ac290458afcb928f68b6cc1c7bdd6963fc72f52f365a2a89c1b5                                                                                                              0.2s
 => [2/5] RUN set -eux;   sed -i 's!http://dl-cdn.alpinelinux.org/!https://alpine.global.ssl.fastly.net/!g' /etc/apk/repositories &&   apk add --no-cache     iproute2     docker-cli     dumb-init    7.5s
 => [3/5] COPY ./*.sh /usr/local/bin/                                                                                                                                                                  0.2s
 => [4/5] COPY ./tests/*.sh /usr/local/bin/tests/                                                                                                                                                      0.1s
 => [5/5] WORKDIR /usr/local/bin                                                                                                                                                                       0.1s
 => exporting to image                                                                                                                                                                                 0.5s
 => => exporting layers                                                                                                                                                                                0.5s
 => => writing image sha256:7349ab5126886ba84770b3cead8c19de5fb897eb6b0a5cd6d7e161db1541dffb                                                                                                           0.0s
```

On master;

Docker CLI hangs;

```bash
DOCKER_BUILDKIT=1 docker build https://github.com/docker/docker-bench-security.git#v1.3.5
[+] Building 8.1s (8/8)
 => CACHED [internal] load git source https://github.com/docker/docker-bench-security.git#v1.3.5                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/alpine:3.10                                                                                                                                         0.0s
 => [1/5] FROM docker.io/library/alpine:3.10                                                                                                                                                           0.0s
 => CACHED [2/5] RUN set -eux;   sed -i 's!http://dl-cdn.alpinelinux.org/!https://alpine.global.ssl.fastly.net/!g' /etc/apk/repositories &&   apk add --no-cache     iproute2     docker-cli     dumb  0.0s
 => CACHED [3/5] COPY ./*.sh /usr/local/bin/                                                                                                                                                           0.0s
 => CACHED [4/5] COPY ./tests/*.sh /usr/local/bin/tests/                                                                                                                                               0.0s
 => CACHED [5/5] WORKDIR /usr/local/bin                                                                                                                                                                0.0s
 => ERROR exporting to image                                                                                                                                                                           0.0s
 => => exporting layers                                                                                                                                                                                0.0s
 => => writing image sha256:accf73094158d5531fa6c29bdf5554f6b7eecbb9864fd6ad5d35eb7e5bae0749                                                                                                           0.0s
 => => naming to                                                                                                                                                                                       8.0s
^CERRO[0008] got 3 SIGTERM/SIGINTs, forcing shutdown
````

With buildx the error also occurs, but it doesn't hang:

```bash
docker buildx version
github.com/docker/buildx v0.8.2 6224def4dd2c3d347eee19db595348c50d7cb491

docker buildx build https://github.com/docker/docker-bench-security.git#v1.3.5
[+] Building 11.8s (8/8) FINISHED
 => [internal] load git source https://github.com/docker/docker-bench-security.git#v1.3.5                                                                                                              2.6s
 => [internal] load metadata for docker.io/library/alpine:3.10                                                                                                                                         0.0s
 => [1/5] FROM docker.io/library/alpine:3.10                                                                                                                                                           0.0s
 => [2/5] RUN set -eux;   sed -i 's!http://dl-cdn.alpinelinux.org/!https://alpine.global.ssl.fastly.net/!g' /etc/apk/repositories &&   apk add --no-cache     iproute2     docker-cli     dumb-init    5.6s
 => [3/5] COPY ./*.sh /usr/local/bin/                                                                                                                                                                  0.1s
 => [4/5] COPY ./tests/*.sh /usr/local/bin/tests/                                                                                                                                                      0.0s
 => [5/5] WORKDIR /usr/local/bin                                                                                                                                                                       0.0s
 => ERROR exporting to image                                                                                                                                                                           0.5s
 => => exporting layers                                                                                                                                                                                0.5s
 => => writing image sha256:619742465254ba9662a5ee8433919f35498bc958af2ae39b76505da63de76dd3                                                                                                           0.0s
 => => naming to                                                                                                                                                                                       3.0s
------
 > exporting to image:
------
error: failed to solve: invalid reference format
```

With BuildKit disabled, it works as expected;

```bash
DOCKER_BUILDKIT=0 docker build https://github.com/docker/docker-bench-security.git#v1.3.5

DOCKER_BUILDKIT=0 docker build https://github.com/docker/docker-bench-security.git#v1.3.5
Sending build context to Docker daemon  136.2kB
Step 1/9 : FROM alpine:3.10
 ---> e7b300aee9f9
Step 2/9 : LABEL   org.label-schema.name="docker-bench-security"   org.label-schema.url="https://dockerbench.com"   org.label-schema.vcs-url="https://github.com/docker/docker-bench-security.git"
 ---> Using cache
 ---> 6a12ab30ca54
Step 3/9 : RUN set -eux;   sed -i 's!http://dl-cdn.alpinelinux.org/!https://alpine.global.ssl.fastly.net/!g' /etc/apk/repositories &&   apk add --no-cache     iproute2     docker-cli     dumb-init
 ---> Using cache
 ---> 11813d801d06
Step 4/9 : COPY ./*.sh /usr/local/bin/
 ---> Using cache
 ---> 1b30d93bf486
Step 5/9 : COPY ./tests/*.sh /usr/local/bin/tests/
 ---> Using cache
 ---> 3adc8d904556
Step 6/9 : HEALTHCHECK CMD exit 0
 ---> Using cache
 ---> 092c4e4e7695
Step 7/9 : WORKDIR /usr/local/bin
 ---> Using cache
 ---> ed88260a96a4
Step 8/9 : ENTRYPOINT [ "/usr/bin/dumb-init", "docker-bench-security.sh" ]
 ---> Using cache
 ---> aa6f3e91bccc
Step 9/9 : CMD [""]
 ---> Using cache
 ---> 709ff8947191
Successfully built 709ff8947191
```

### PR DESCRIPTION
**Pull request by crazy-max**
fixes #43653 

fix image reference not checked properly.

cc @thaJeztah 

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: builder/builder-next/exporter/export.go [hunk 1]
```diff
@@ -196,14 +198,13 @@ func (e *imageExporterInstance) Export(ctx context.Context, inp exporter.Source,
 	}
 	_ = configDone(nil)
 
-	if e.opt.ReferenceStore != nil {
-		targetNames := strings.Split(e.targetName, ",")
-		for _, targetName := range targetNames {
-			tagDone := oneOffProgress(ctx, "naming to "+targetName)
+	if e.opt.ReferenceStore != nil && e.targetName != "" {
+		for _, targetName := range strings.Split(e.targetName, ",") {
 			tref, err := distref.ParseNormalizedNamed(targetName)
```
Comment (thaJeztah) [inline builder/builder-next/exporter/export.go:203]: Do we need a check inside the loop to check for;

- `targetName` being empty
- `targetName` being a whitespace-only string?


Thinking of cases where (idk) something like `"   , , , "` would ever find it's way here

Comment (thaJeztah) [reply builder/builder-next/exporter/export.go:203]: So something like;

```go
targetName = strings.TrimSpace(targetName)
if targetName == "" {
	continue
}
```

**Review by thaJeztah — APPROVED** [review]
LGTM


for others; the CI in moby does not use buildkit currently, so if we want a test for this, that can be added in the BuildKit repository (from which we run e2e tests)

**Comment by thaJeztah** [comment]
Failure is unrelated; known flaky test; https://github.com/moby/moby/issues/42698
WIP fix is in https://github.com/moby/moby/pull/42829

```
=== FAIL: libnetwork/networkdb TestNetworkDBNodeJoinLeaveIteration (1.06s)
[2022-05-28T16:21:58.000Z]     networkdb_test.go:535: The networkNodes list has to have be 2 instead of 1 - [5f7df566397e]
```

**Review by rumpl — APPROVED** [review]

**Comment by thaJeztah** [comment]
Having a chat with Tõnis, and looking at the root cause; looks to be introduced in https://github.com/moby/moby/commit/a2aaf4cc83aa88aa09bb1d4bc88f36bf68cccd66#diff-b10bf49d6dea68850c31f77500558d1b3a49c1909992c3dced6ec8eb0cf6e383L72


And the alternative is to revert some of those changes.

I'm personally ok with taking this PR as-is (but considering it a temporary fix) and do the better/cleaner approach after this

**Comment by crazy-max** [comment]
> Having a chat with Tõnis, and looking at the root cause; looks to be introduced in https://github.com/moby/moby/commit/a2aaf4cc83aa88aa09bb1d4bc88f36bf68cccd66#diff-b10bf49d6dea68850c31f77500558d1b3a49c1909992c3dced6ec8eb0cf6e383L72

Yes indeed, previous logic is better so it errored early and does not wait for the build to complete. Not sure why I changed that for BuildKit 0.10 :thinking_face:. Let me know if it lgty @tonistiigi.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks for updating

**Review by tonistiigi — APPROVED** [review]

**Comment by thaJeztah** [comment]
Windows failure is unrelated (know  flaky test)

```
=== RUN   TestDockerSuite/TestRunInteractiveWithRestartPolicy
    docker_cli_run_test.go:1795: assertion failed: 
        Command:  d:\CI\PR-43654\3\binary\docker.exe run -i --name test-inter-restart --restart=always busybox sh
        ExitCode: 0
        Stdout:   
        Stderr:   
        
        Failures:
        ExitCode was 0 expected 11
    --- FAIL: TestDockerSuite/TestRunInteractiveWithRestartPolicy (6.36s)

**Comment by thaJeztah** [comment]
everything else passed; bringing this one in; thanks!

**Pull request merged**